File: Sources/WattsonCore/Core/Collector.swift
// Core/Collector.swift — 驱动解析管线产出 Snapshot（rows + 新鲜度拆分）。
// parseAll / loadCache 可注入（测试缝）；生产路径用 SessionScanner 全量扫描 +
// 内存缓存。任何刷新失败都落进 snapshot.errors，旧行原样保留；失败轮只推进
// fetchedAt（尝试时间），lastSuccessAt 保留上一个成功值。
import Foundation

public struct Snapshot: Sendable, Equatable {
    public var rows: [UsageRow]
    public var fetchedAt: Double
    /// 最近一次「成功」采集时间；失败轮保留旧值
    public var lastSuccessAt: Double?
    public var errors: [String]
    public var refreshing: Bool
    /// 聚合响应体缓存以「行数组整体替换」为失效边界：每次 refresh 产生新实例身份
    public let instanceID = UUID()

    public init(rows: [UsageRow] = [], fetchedAt: Double = 0, lastSuccessAt: Double? = nil,
                errors: [String] = [], refreshing: Bool = false) {
        self.rows = rows
        self.fetchedAt = fetchedAt
        self.lastSuccessAt = lastSuccessAt
        self.errors = errors
        self.refreshing = refreshing
    }
}

func errMsg(_ err: Error) -> String { errText(err) }

public final class Collector: @unchecked Sendable {
    public let devices: [DeviceRoot]
    public let localHost: String
    private let lock = NSLock()
    private var state: Snapshot
    private var inflight: Task<Snapshot, Never>?
    private let parseAll: @Sendable () async throws -> Void
    private let loadCacheOpt: @Sendable () async throws -> SessionCache

    public init(devices: [DeviceRoot]? = nil, localHost: String? = nil,
                parseAll: (@Sendable () async throws -> Void)? = nil,
                loadCache: (@Sendable () async throws -> SessionCache)? = nil) {
        self.devices = devices ?? loadDeviceRoots()
        self.localHost = localHost ?? shortHostname()
        self.state = Snapshot()
        self.parseAll = parseAll ?? { try await parseAllSessions() }
        self.loadCacheOpt = loadCache ?? { try loadSessionCache() }
    }

    public var snapshot: Snapshot {
        lock.lock(); defer { lock.unlock() }
        return state
    }

    /// 并发/排队语义：进行中的刷新不重复触发，后来者共享同一轮结果；
    /// 刷新结束后再来的调用开启新一轮。
    public func refresh() async -> Snapshot {
        lock.lock()
        if let t = inflight {
            lock.unlock()
            return await t.value
        }
        var start = state
        start.refreshing = true
        state = start
        let task = Task<Snapshot, Never> { await self.performRefresh(start) }
        inflight = task
        lock.unlock()
        let v = await task.value
        return v
    }

    private func performRefresh(_ start: Snapshot) async -> Snapshot {
        var errors: [String] = []
        do {
            try await parseAll()
        } catch {
            errors.append("parseAllSessions: \(errMsg(error))")
        }
        var rows = start.rows
        if errors.isEmpty {
            do {
                let cache = try await loadCacheOpt()
                rows = rowsFromCache(cache, devices, localHost)
            } catch {
                errors.append("loadCache: \(errMsg(error))")
            }
        }
        let now = Date.nowMs()
        let settled = Snapshot(
            rows: rows,
            fetchedAt: now,
            lastSuccessAt: errors.isEmpty ? now : start.lastSuccessAt,
            errors: errors,
            refreshing: false)
        lock.lock()
        self.state = settled
        self.inflight = nil
        lock.unlock()
        return settled
    }
}

File: server/src/collector.ts
// server/src/collector.ts — builds a Snapshot of usage rows by driving the
// codeburn parse pipeline (parseAllSessions → session cache → rowsFromCache).
//
// Deviations from the brief sketch (documented in task-7-report.md):
// 1. `parseAllSessions` / `loadCache` are imported *lazily* (dynamic import at
//    refresh time) instead of at module top level. `@codeburn/parser.js` pulls
//    in every provider module (~30 eager imports); keeping it lazy lets tests
//    import this module cheaply and defers the heavy graph until first refresh.
// 2. Both are injectable via constructor opts (test seam): `parseAll` /
//    `loadCache`. Production passes nothing and gets the real codeburn fns.
// 3. `rowsFromCache` runs inside the same try/catch as `loadCache` so ANY
//    refresh failure lands in snapshot.errors and old rows are preserved
//    (the brief's stated contract; the sketch left it outside the try).
import { hostname } from 'node:os'
import { loadDeviceRoots, type DeviceRoot } from '@codeburn/providers/device-roots.js'
import { rowsFromCache, type UsageRow } from './aggregate.js'
import type { SessionCache } from '@codeburn/session-cache.js'

export type Snapshot = {
  rows: UsageRow[]
  fetchedAt: number
  // 最近一次「成功」采集时间；失败轮保留旧值。fetchedAt 是最后一次尝试时间——
  // 失败时也推进，若不拆分，「数据时间」会把失败尝试误报成新数据
  lastSuccessAt: number | null
  errors: string[]
  refreshing: boolean
}

const errMsg = (err: unknown): string => (err instanceof Error ? err.message : String(err))

export class Collector {
  readonly devices: DeviceRoot[]
  readonly localHost: string
  snapshot: Snapshot = { rows: [], fetchedAt: 0, lastSuccessAt: null, errors: [], refreshing: false }
  private lock: Promise<void> | null = null
  private readonly parseAll?: () => Promise<unknown>
  private readonly loadCacheOpt?: () => Promise<SessionCache>

  constructor(opts: { devices?: DeviceRoot[]; localHost?: string; parseAll?: () => Promise<unknown>; loadCache?: () => Promise<SessionCache> } = {}) {
    this.devices = opts.devices ?? loadDeviceRoots()
    this.localHost = opts.localHost ?? hostname().replace(/\..*$/, '')
    this.parseAll = opts.parseAll
    this.loadCacheOpt = opts.loadCache
  }

  async refresh(): Promise<Snapshot> {
    if (this.lock) return this.lock.then(() => this.snapshot)
    this.snapshot = { ...this.snapshot, refreshing: true }
    this.lock = (async () => {
      const errors: string[] = []
      try {
        if (this.parseAll) await this.parseAll()
        else await (await import('@codeburn/parser.js')).parseAllSessions() // 全量管线，增量缓存使其快速；同时把缓存写盘
      } catch (err) { errors.push(`parseAllSessions: ${errMsg(err)}`) }
      let rows: UsageRow[] = this.snapshot.rows
      if (errors.length === 0) {
        try {
          const cache = this.loadCacheOpt
            ? await this.loadCacheOpt()
            : await (await import('@codeburn/session-cache.js')).loadCache()
          rows = rowsFromCache(cache, this.devices, this.localHost)
        } catch (err) { errors.push(`loadCache: ${errMsg(err)}`) }
      }
      // 失败轮：fetchedAt 推进（尝试时间），lastSuccessAt 保留上一个成功值
      this.snapshot = {
        rows,
        fetchedAt: Date.now(),
        lastSuccessAt: errors.length === 0 ? Date.now() : this.snapshot.lastSuccessAt,
        errors,
        refreshing: false,
      }
    })().finally(() => { this.lock = null })
    return this.lock.then(() => this.snapshot)
  }
}
